### PR TITLE
Fog of war mechanic (Bandit)

### PR DESCRIPTION
**Pull request by zomglings**
This pull request introduces the single player fog of war mechanic that we will be using in The Degen Trail.

### CODE REVIEW

File: src/Bandit.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    // Block number for last roll for an NFT.
+    mapping(address => mapping(uint256 => uint256)) public LastRollForNFT;
+
+    error PlayerDeadlineExceeded(address player);
+    error NFTDeadlineExceeded(address tokenAddress, uint256 tokenID);
+    error WaitForPlayerTick(address player);
+    error WaitForNFTTick(address tokenAddress, uint256 tokenID);
+    error NFTNotOwnedByPlayer(address player, address tokenAddress, uint256 tokenID);
+
+    function _preRollForNFT(address tokenAddress, uint256 tokenID) internal virtual {}
```
Comment (kellan-simiotics) [inline src/Bandit.sol:61]: Should there also be a `preRollForPlayer` ?

File: test/Bandit.t.sol [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        nfts.mint(player1, tokenID);
+        feeToken.mint(player1, rollFee + rerollFee);
+        feeToken.approve(address(bandit), rollFee + rerollFee);
+        vm.expectEmit(true, true, true, false, address(bandit));
+        emit NFTRoll(address(nfts), tokenID);
+        bandit.rollForNFT(address(nfts), tokenID);
+        vm.roll(block.number + blockDeadline + 1);
+        vm.expectRevert(abi.encodeWithSelector(Bandit.NFTDeadlineExceeded.selector, address(nfts), tokenID));
+        bandit.resolveForNFT(address(nfts), tokenID);
+        vm.stopPrank();
```
Comment (kellan-simiotics) [inline test/Bandit.t.sol:261]: Is this test supposed to `rerollForNFT` ?

**Review by kellan-simiotics — CHANGES_REQUESTED** [review]

File: src/Bandit.sol (hunk 1 above)
Comment (zomglings) [reply src/Bandit.sol:61]: Added a comment clarifying why there isn't symmetry between player and NFT - NFT has preroll, but player doesn't.

File: test/Bandit.t.sol (hunk 2 above)
Comment (zomglings) [reply test/Bandit.t.sol:261]: Thank you, fixed.

**Review by kellan-simiotics — APPROVED** [review]
lg

**Pull request merged**